pkg/daemons/executor/embed_windows.go

```diff
@@ -0,0 +1,89 @@
+//go:build windows && !no_embedded_executor
+// +build windows,!no_embedded_executor
+
+package executor
+
+import (
+	"encoding/json"
+	"os"
+	"os/exec"
+	"strings"
+	"time"
+
+	"github.com/Microsoft/hcsshim"
+	"github.com/sirupsen/logrus"
+
+	// registering k3s cloud provider
+	_ "github.com/k3s-io/k3s/pkg/cloudprovider"
+	daemonconfig "github.com/k3s-io/k3s/pkg/daemons/config"
+)
+
+const (
+	networkName = "flannel.4096"
+)
+
+type SourceVipResponse struct {
+	IP4 struct {
+		IP string `json:"ip"`
+	} `json:"ip4"`
+}
+
+func platformKubeProxyArgs(nodeConfig *daemonconfig.Node) map[string]string {
+	argsMap := map[string]string{}
+	argsMap["network-name"] = networkName
+	if sourceVip := waitForSourceVip(networkName, nodeConfig); sourceVip != "" {
+		argsMap["source-vip"] = sourceVip
+	}
+	return argsMap
+}
+
+func waitForSourceVip(networkName string, nodeConfig *daemonconfig.Node) string {
+	for range time.Tick(time.Second * 5) {
+		network, err := hcsshim.GetHNSNetworkByName(networkName)
+		if err != nil {
+			logrus.WithError(err).Warningf("can't find HNS network, retrying %s", networkName)
+			continue
+		}
+		if network.ManagementIP == "" {
+			continue
+		}
+
+		subnet := network.Subnets[0].AddressPrefix
+
+		configData := `{
+			"cniVersion": "0.2.0",
+			"name": "vxlan0",
+			"ipam": {
+				"type": "host-local",
+				"ranges": [[{"subnet":"` + subnet + `"}]],
+				"dataDir": "/var/lib/cni/networks"
+			}
+		}`
+
+		cmd := exec.Command("host-local.exe")
+		cmd.Env = append(os.Environ(),
+			"CNI_COMMAND=ADD",
+			"CNI_CONTAINERID=dummy",
+			"CNI_NETNS=dummy",
+			"CNI_IFNAME=dummy",
+			"CNI_PATH="+nodeConfig.AgentConfig.CNIBinDir,
+		)
+
+		cmd.Stdin = strings.NewReader(configData)
+		out, err := cmd.Output()
```

Comment:
I am confused about this code. You are fetching the subnet of the flannel.4096 HNS overlay network that was created in the agent/flannel code. That subnet will be the podCIDR assigned to the node, right? Then you use host-local to "reserve" that subnet, right? I don't understand why do you need to do that for kube-proxy in Windows, why you need the response from the CNI and what do you mean by sourceVIP. Could you explain briefly? Thanks a lot!

Reply:
@manuelbuil please excuse me I am still navigating inside Microsoft for who the best to answer to this question. And just in case you are interested, here is the original pull request when the sourceVIP concept was first introduced: https://github.com/kubernetes/kubernetes/pull/70896

Reply:
Thanks for the link, I think I understand now why a source-vip is needed. In Linux we are using DSR and the source IP never changes, so we don't need such source-vip. According to the PR that you linked, kube-proxy in windows could work in DSR or non-DSR mode. I understand you are only implementing the non-DSR option, right?

I still don't understand why you need to create a "dummy host-local network". If you could find that information, that 'd be great, thanks

Reply:
yes, this was only implemented with non-DSR mode. I will need to check the Windows support of DSR mode (I knew there is a feature gate WinDSR, but I don't know what's the support status for that.)

And for `host-local`, we are reusing this IPAM plugin to help allocate an IP address from the subnet of the underlying network. The underlying HNS network was already created when flannel starts (see cniConf in pkg/agent/flannel/setup_windows.go). Then at the moment to start kube-proxy, we grab & pass the subnet information to `host-local` plugin to allocate an IP address valid in the subnet as the source VIP, and in turn we pass it into `kube-proxy`. 

So, no `dummy` network was created at this point. (We added those `dummy` to make the `host-local` happy since it checks all the parameters existence.)

Reply:
Ok, got it, thanks